### PR TITLE
Improve build isolation in `PythonPipBuilder`

### PR DESCRIPTION
**Pull request by haampie**
This PR supersedes and is the more general / proper fix for #31131

Needs #40234 (currently rebased on it).

We run pip with `--no-build-isolation` because we don't wanna let pip install
build deps. As a consequence, when pip runs hooks, it runs hooks of *any*
package it can find in `sys.path` [^1]

This is problematic both for external and non-external python.

- For external python: pip may run hooks of any package in system and user
  site-packages dirs
- For non-external python: spack by design doesn't have system site-packages
  [^2], or at least that's empty, but user site-packages are still picked up.

An example of the problem is when you e.g. `spack install py-setuptools@63 ^python@3.9`
when you have e.g. `setuptools==44.0.0` installed as a user package in
~/.local.

During the build,  `pip` executes the hook `setuptools.finalize_distribution_options`,
and looks for entry points through (vendored) `importlib.metadata.entry_points(...)`.
This finds `~/.local/.../setuptools.egg-info/entry_points.txt` and executes its
hooks, and that fails badly, cause `setuptools` that's being built is already
imported, and the old hooks of user setuptools refer to functions that no
longer exist in newer setuptools.

#### Solution for external Python

Creating an empty venv before invoking pip in Spack's build env ensures that
pip can no longer see user/system site-package, but is limited to standard
libraries and packages in `PYTHONPATH`.

The downside of this approach is that shebangs generated by pip now will point
to the temporary virtual environment path, instead of the underlying python.
The current workaround for that is to fixup shebangs post install.

#### Solution for Spack managed Python

We only have to drop _user_ site-packages from `sys.path`, and for that an
actual environment variable exists: `PYTHONNOUSERSITE=1`, so pip is invoked
with that.

#### Alternative solutions

Alternative solution that didn't work well was to try and run pip like `python -S -m pip`,
which also drops the system / user site packages paths -- problem is that pip
spawns subprocesses and those flags don't propagate.

Using environment variables for external Python is also not an option, because
there's no corresponding env variable to `-S`, only to `-s`.

TL;DR: use venv so pip cannot run hooks from system / user packages, only from
stuff in PYTHONPATH, aka our "build environment".

[^1]: This is through `importlib.metadata.entry_points(...)`; for older Python
and older setuptools things are even worse, cause it has a buggy old vendored
importlib.
[^2]: although users may install py-pip ^python, and then "accidentally"
pollute the original spack python prefix with pip install system packages.

### CODE REVIEW

**Comment by haampie** [comment]
Dunno if somebody can suggest a better solution for ensuring `console_script` entrypoints/script have shebangs point to the underlying python instead of the temporary build venv path.

**Comment by haampie** [comment]
@adamjstewart thoughts?

This works in Gitlab. The GHA failure is also correct (setuptools is missing as a dependency in py-isort). Fixed in #40234.

The fact that the GHA failed with this PR only shows that we need this for better isolation from system python packages :)

**Comment by haampie** [comment]
Alternative / compromise:

1. If Python is non-external, it should suffice to run pip with `PYTHONNOUSERSITE=1`, since system site-package is empty.
2. If Python is external, create a venv and patch shebangs post install.

**Comment by johnwparent** [comment]
> Alternative / compromise:
> 
> 1. If Python is non-external, it should suffice to run pip with `PYTHONNOUSERSITE=1`, since system site-package is empty.
> 2. If Python is external, create a venv and patch shebangs post install.

Would using or extending [EnvBuilder](https://docs.python.org/3/library/venv.html#venv.EnvBuilder) work for this case? At a glance it looks like [this method](https://docs.python.org/3/library/venv.html#venv.EnvBuilder.install_scripts) could be leveraged to do what you want.

**Comment by haampie** [comment]
> > Alternative / compromise:
> > 
> > 1. If Python is non-external, it should suffice to run pip with `PYTHONNOUSERSITE=1`, since system site-package is empty.
> > 2. If Python is external, create a venv and patch shebangs post install.
> 
> Would using or extending [EnvBuilder](https://docs.python.org/3/library/venv.html#venv.EnvBuilder) work for this case? At a glance it looks like [this method](https://docs.python.org/3/library/venv.html#venv.EnvBuilder.install_scripts) could be leveraged to do what you want.

Not immediately sure if that can be used.

Bit more context, [here's an example `pyproject.toml`](https://github.com/psf/black/blob/9b82120ddb81373377b58da5de7caa9495a1551e/pyproject.toml#L86-L88) that lets pip generate two wrapper scripts `<pkg>/bin/black` and `<pkg>bin/blackd`.

Here's where pip figures out the shebang line:

https://github.com/pypa/pip/blob/2ba5acc8a4772ca1dc0d62a2dfe8967af28649d6/src/pip/_vendor/distlib/scripts.py#L166

The `get_executable` returns `sys.executable`, and this points to the virtual environment's python. Can't and shouldn't change this, cause `sys.executable` is also used to spawn subprocesses for which we do want to use the venv python s.t. system/user search paths are disabled.

The  only way to override this, is if you can somehow pass a custom executable [here](https://github.com/pypa/pip/blob/2ba5acc8a4772ca1dc0d62a2dfe8967af28649d6/src/pip/_vendor/distlib/scripts.py#L162-L162) or [here](https://github.com/pypa/pip/blob/2ba5acc8a4772ca1dc0d62a2dfe8967af28649d6/src/pip/_vendor/distlib/scripts.py#L181-L181), but that looks impossible. (especially since all we have is the pip command line...)

**Comment by haampie** [comment]
@pradyunsg, do you have opinions about this? The top level comment was updated to describe the issue and solution, the rest of the comments is not relevant.

**Comment by pradyunsg** [comment]
> problem is that pip
> spawns subprocesses and those flags don't propagate.

Hmm... we could make pip respect this in a future release, I think. File an issue over at pypa/pip about this?

This might also get resolved once we get to the possible-but-needs-work improvement to actually avoid pip-subprocesses inside pip. :)

---

This seems reasonable to me, with the caveat that it's... a few layers of detail-dependent bodges/behaviours (which tends toward some fragility), but they're reasonable to solve the problem as long as that fragility isn't a huge concern.

**Comment by adamjstewart** [comment]
Fantastic work, completely agree with all of the decisions reached here. I like the path forward of:

1. Use `PYTHONNOUSERSITE=1` for internal Python
2. Use a venv for external Python
3. Open an issue in pip's repo about subprocesses
4. Remove the venv stuff in the distant future

The fact that this PR discovered bugs in our recipes is all the proof I need that this is useful. I'll review any separate PRs for bugs found in packages and then we can merge this. Thanks for reviewing @pradyunsg.

TIL about footnotes![^1]

[^1]: https://github.blog/changelog/2021-09-30-footnotes-now-supported-in-markdown-fields/!

**Comment by haampie** [comment]
I still have to check if I need to add more patching: https://github.com/pypa/pip/issues/12309#issuecomment-1744713899

Edit: ok, everything seems to be about shebangs really, and I'm recursing the full install prefix, so should be fine.

**Comment by haampie** [comment]
@adamjstewart should be ready for review now

**Comment by haampie** [comment]
And by now I mean now.

Coverage is low, but it's tested in bootstrapping gha w/o coverage anyways.

File: lib/spack/spack/build_systems/python.py [hunk 1]
```diff
@@ -447,8 +494,36 @@ def global_options(self, spec, prefix):
...
+    def _build_venv_path(self):
+        """Return the path to the virtual environment used for building when
+        python is external."""
+        return os.path.join(self.spec.package.stage.path, "build_env")
+
+    @property
+    def _build_venv_python(self) -> Executable:
+        """Return the Python executable in the build virtual environment when
+        python is external."""
+        return Executable(os.path.join(self._build_venv_path, "bin", "python"))
```
Comment (adamjstewart) [inline lib/spack/spack/build_systems/python.py:507]: Will it always be `python`? For example, on Ubuntu where only `python3` exists.

Comment (haampie) [reply lib/spack/spack/build_systems/python.py:507]: Since Python 2.7 apparently: https://github.com/python/cpython/commit/7ded1f0f694f0f99252ea19eca18b74ea5e36cb0

**Review by adamjstewart — APPROVED** [review]

**Comment by haampie** [comment]
Thanks for the review all :)
